substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,420 @@
+
+		// a validator is also a nominator.
+		Self::on_nominator_add(who)
+	}
+
+	// Fired when someone removes their intention to nominate, either due to chill or validating.
+	//
+	// Note: it is assumed that who's staking state is updated *before* the caller calling into
+	// this method. Thus, the nominations before the nominator has been removed from staking are
+	// passed in, so that the target list can be updated accordingly.
+	fn on_nominator_remove(who: &T::AccountId, nominations: Vec<T::AccountId>) {
+		let nominator_vote = Self::weight_of(Self::active_vote_of(who));
+
+		log!(
+			debug,
+			"on_nominator_remove: {:?} with {:?}. impacting {:?}",
+			who,
+			nominator_vote,
+			nominations,
+		);
+
+		// updates the nominated target's score. Note: this may update the score of up to
+		// `T::MaxNominations` validators.
+		for t in nominations.iter() {
+			Self::update_score::<T::TargetList>(&t, StakeImbalance::Negative(nominator_vote.into()))
+		}
+
+		let _ = T::VoterList::on_remove(&who).defensive_proof(
+			"the nominator exists in the list as per the contract with staking; qed.",
+		);
+	}
+
+	// Fired when someone removes their intention to validate, either due to chill or nominating.
+	fn on_validator_remove(who: &T::AccountId) {
+		log!(debug, "on_validator_remove: {:?}", who,);
+
+		let _ = T::TargetList::on_remove(&who).defensive_proof(
+			"the validator exists in the list as per the contract with staking; qed.",
+		);
+
+		// validator is also a nominator.
+		Self::on_nominator_remove(who, vec![]);
+	}
+
+	// Fired when an existing nominator updates their nominations.
+	//
+	// This is called when a nominator updates their nominations. The nominator's stake remains the
+	// same (updates to the nominator's stake should emit [`Self::on_stake_update`] instead).
+	// However, the score of the nominated targets must be updated accordingly.
+	//
+	// Note: it is assumed that who's staking state is updated *before* calling this method.
+	fn on_nominator_update(who: &T::AccountId, prev_nominations: Vec<T::AccountId>) {
+		let nominator_vote = Self::weight_of(Self::active_vote_of(who));
+
+		let curr_nominations =
+			<T::Staking as StakingInterface>::nominations(&who).unwrap_or_default();
+
+		log!(
+			debug,
+			"on_nominator_update: {:?}, with {:?}. previous nominations: {:?} -> new nominations {:?}",
+			who, nominator_vote, prev_nominations, curr_nominations,
+		);
+
+		// new nominations
+		for target in curr_nominations.iter() {
+			if !prev_nominations.contains(target) {
+				Self::update_score::<T::TargetList>(
+					&target,
+					StakeImbalance::Positive(nominator_vote.into()),
+				);
+			}
+		}
+		// removed nominations
+		for target in prev_nominations.iter() {
+			if !curr_nominations.contains(target) {
+				Self::update_score::<T::TargetList>(
+					&target,
+					StakeImbalance::Negative(nominator_vote.into()),
+				);
+			}
+		}
+	}
```

Comment:
`O(2(n*m))`, but should be fine because there's a tight cap on how many nominations there can be, right? Maybe add a note explaining that?